roofit/roofitcore/inc/RooMinimizer.h

```diff
@@ -100,37 +117,103 @@ class RooMinimizer : public TObject {
   void profileStop() ;
 
-  inline Int_t getNPar() const { return fitterFcn()->NDim() ; }
+  inline Int_t getNPar() const { return fitterFcn()->getNDim() ; }
   inline std::ofstream* logfile() { return fitterFcn()->GetLogFile(); }
   inline Double_t& maxFCN() { return fitterFcn()->GetMaxFCN() ; }
-  
-  const RooMinimizerFcn* fitterFcn() const {  return ( fitter()->GetFCN() ? ((RooMinimizerFcn*) fitter()->GetFCN()) : _fcn ) ; }
-  RooMinimizerFcn* fitterFcn() { return ( fitter()->GetFCN() ? ((RooMinimizerFcn*) fitter()->GetFCN()) : _fcn ) ; }
+
+  const RooAbsMinimizerFcn *fitterFcn() const;
+  RooAbsMinimizerFcn *fitterFcn();
+
+  bool fitFcn() const;
 
 private:
+  template <typename MinimizerFcn = RooMinimizerFcn>
+  RooMinimizer(RooAbsReal &function, MinimizerFcn* /* used only for template deduction */);
 
-  Int_t       _printLevel ;
-  Int_t       _status ;
-  Bool_t      _profile ;
+  Int_t _printLevel = 1;
+  Int_t _status = -99;
+  Bool_t _profile = kFALSE;
 
-  Bool_t      _verbose ;
-  TStopwatch  _timer ;
-  TStopwatch  _cumulTimer ;
-  Bool_t      _profileStart ;
+  Bool_t _verbose = kFALSE;
+  TStopwatch _timer;
+  TStopwatch _cumulTimer;
+  Bool_t _profileStart = kFALSE;
 
-  TMatrixDSym* _extV ;
+  TMatrixDSym *_extV = 0;
 
-  RooMinimizerFcn *_fcn;
-  std::string _minimizerType;
+  RooAbsMinimizerFcn *_fcn;
+  std::string _minimizerType = "Minuit2";
+  FcnMode _fcnMode;
 
   static ROOT::Fit::Fitter *_theFitter ;
 
   std::vector<std::pair<std::string,int> > _statusHistory ;
 
   RooMinimizer(const RooMinimizer&) ;
-	
+
   ClassDef(RooMinimizer,1) // RooFit interface to ROOT::Fit::Fitter
 } ;
 
+////////////////////////////////////////////////////////////////////////////////
+/// Construct MINUIT interface to given function. Function can be anything,
+/// but is typically a -log(likelihood) implemented by RooNLLVar or a chi^2
+/// (implemented by RooChi2Var). Other frequent use cases are a RooAddition
+/// of a RooNLLVar plus a penalty or constraint term. This class propagates
+/// all RooFit information (floating parameters, their values and errors)
+/// to MINUIT before each MINUIT call and propagates all MINUIT information
+/// back to the RooFit object at the end of each call (updated parameter
+/// values, their (asymmetric errors) etc. The default MINUIT error level
+/// for HESSE and MINOS error analysis is taken from the defaultErrorLevel()
+/// value of the input function.
+
+template <typename MinimizerFcn>
+RooMinimizer::RooMinimizer(RooAbsReal &function, MinimizerFcn* /* value unused */)
+{
+   RooSentinel::activate();
+
+   if (_theFitter)
+      delete _theFitter;
+   _theFitter = new ROOT::Fit::Fitter;
+   _theFitter->Config().SetMinimizer(_minimizerType.c_str());
+   setEps(1.0); // default tolerance
+
+   _fcn = new MinimizerFcn(&function, this, _verbose);
+
+   // make sure to order correctly so that child classes get checked for first TODO
+   if (dynamic_cast<RooGradMinimizerFcn *>(_fcn)) {
```

Comment:
```suggestion
   if (std::is_same<MinimizerFcn, RooGradMinimizerFcn>::value) {
```
No need for dynamic cast in this function, because you know the actual type from the template argument above.

Reply:
I'm going to detemplatize the whole thing, so then we can instead just use the FcnMode parameter.